### ISSUE TITLE
Listen to Client whist in headless Mode / showUI:false

### ISSUE THREAD

**Issue by Burnsy** [MagicLabs/magic-js#93]
### ✅ Prerequisites

- [x] Did you perform a cursory search of open issues? Is this feature already requested elsewhere?
- [x] Are you reporting to the correct repository (`magic-sdk`)?

### ✨ Feature Request

When using `loginWithMagicLink({showUI: false})` have a method of monitoring  the state of the client so a custom UI could be implemented with the same logic as the MagicLinks Clients

## 🧩 Context
My thoughts for this is that in some use cases users may get confused as that it may be easier for a developer to build a custom interface instead of the one provided by magic-sdk. 

Currently i believe there is no way to watch the client to give return the states so that could be implemented into a custom UI. 
- Email DNS Warning
- Email Sent to email Address
-  Waiting for the client to accept the link

**Comment by smithki** [comment]
What if we return promievents from the request, instead of plain promises? Here’s an example implementation from [Web3](https://github.com/ethereum/web3.js/tree/1.x/packages/web3-core-promievent).

We could improve on that format with strong typing on a per-method basis, too.

**Comment by Burnsy** [comment]
I think the .on() implementation could really work as you could change a ui state using a hook for example.

**Comment by smithki** [comment]
Here's my take on a `PromiEvent` utility, which I think would meet this use-case quite well:

```
import { EventEmitter } from 'events';
import TypedEmitter from 'typed-emitter';
import { MagicSDKError, MagicRPCError } from '../core/sdk-exceptions';

export type EventsDefinition<EventName extends string = string> = {
  [P in EventName]: (...args: any[]) => void;
};

export type PromiEvent<TResult, TEvents extends EventsDefinition> = Promise<TResult> & TypedEmitter<TEvents>;

type DefaultEvents<TResult> = {
  done: (result: TResult) => void;
  error: (err: MagicSDKError | MagicRPCError) => void;
  settled: () => void;
};

type PromiseExecutor<TResult> = (
  resolve: (value?: TResult | PromiseLike<TResult>) => void,
  reject: (reason?: any) => void,
) => void;

/**
 * Create a native JavaScript `Promise` overloaded with strongly-typed methods
 * from `EventEmitter`.
 */
export function createPromiEvent<TResult, TEvents extends EventsDefinition = {}>(
  executor: PromiseExecutor<TResult>,
): PromiEvent<TResult, TEvents & DefaultEvents<TResult>> {
  const promise = new Promise(executor);
  const eventEmitter = new EventEmitter() as TypedEmitter<TEvents & DefaultEvents<TResult>>;
  const promiEvent = Object.assign(promise, eventEmitter);

  promiEvent.then(result => {
    promiEvent.emit('done', result);
    if (!promiEvent.finally) promiEvent.emit('settled');
  });

  promiEvent.catch(err => {
    promiEvent.emit('error', err);
    if (!promiEvent.finally) promiEvent.emit('settled');
  });

  if (promiEvent.finally) promiEvent.finally(() => promiEvent.emit('settled'));

  return promiEvent;
}
```

Let me know what you think @Burnsy -- I would appreciate feedback.

Also, in the interest of transparency, this sort of change may require some internal discussion here at Magic to ensure it can be implemented from the `<iframe>` side without complication. I'm feeling imaginative this weekend so I cooked this up pretty quickly, but this feature would require some not-insignificant end-to-end testing.

**Comment by Burnsy** [comment]
I am not too familiar on promievents, but it seems Good. I understand there is has to be testing on such changes. 

How many events do you think they would be so that the developer could replicates the states of the current ui!

**Comment by smithki** [comment]
That's a good question, maybe you can help me answer that.

Let's start with the few you mentioned:

- Email DNS Warning (do you mind expanding on this? What are you looking for specifically?)
- Email Sent to email Address
- ~Waiting for the client to accept the link~ I think this is implied if the promise has started

I'm also thinking to have these default events for every request:

- `done`: equal to `Promise.then`
- `error`: equal to `Promise.catch`
- `settled`: equal to `Promise.finally`

BTW, I think most of the same UI states can be replicated with appropriate error handling. [See the error codes here to better understand(https://docs.magic.link/client-sdk/browser-js/sdk/errors-and-warnings#magic-link-error-codes). I am hoping to produce an example soon to demonstrate a white-labeled implementation step-by-step.

In lieu of events for the time being, you can also simply `catch` the request Promise and check for the appropriate error code using the `SDKErrorCode` or `RPCErrorCode` enum.

**Comment by Burnsy** [comment]
This what i meant by the DNS ERROR.

![image](https://user-images.githubusercontent.com/15834048/82762992-33869980-9dfc-11ea-962c-70e873276e31.png)

Looking at the error codes i guess its MagicLinkFailedVerification

**Comment by smithki** [comment]
We are moving forward with the `PromiEvent` pattern to solve this problem. These are the initial event types we are interested in implementing:

**Login Flow**
- `email-not-deliverable`: Dispatched if the magic link email is unable to be delivered.

- `email-sent`: Dispatched when the magic link email has been successfully sent from the Magic Link server.

- `retry`: Dispatched when the user restarts the flow. This can only happen if `showUI: true`.

**Update Email Flow**
- `new-email-confirmed`: Dispatched when the magic link has been clicked from the user’s new email address.

- `email-sent`:  Dispatched when the magic link email has been successfully sent from the Magic Link server to the user’s new email address.

- `email-not-deliverable`: Dispatched if the magic link email is unable to be delivered to the user’s new email address.

- `old-email-confirmed`: Dispatched when the magic link has been clicked from the user’s previous email address.

- `retry`: Dispatched when the user restarts the flow. This can only happen if `showUI: true`.

---

**NOTE:** Any of the above events are liable to dispatch more than once if `showUI: true`. Plan to handle accordingly in client-side code 😃

**Comment by smithki** [comment]
The initial implementation landed in `v1.4.0`. Documentation will be added soon (tracked by #98)
